File: src/implementation/writer/writer.zig
const std = @import("std");
const internal = @import("../../internal.zig");
const lib = internal.lib;

pub const InterfaceWriter = struct {
    writer: lib.GciInterfaceWriter,

    pub fn write(writer: InterfaceWriter, data: []const u8) !void {
        const result = lib.gci_writer_write(writer.writer, data.ptr, data.len);
        if (result != data.len) {
            return error.Writer;
        }
    }
};

inline fn writeData(writer: *const anyopaque, data: []const u8) !void {
    const result = lib.gci_writer_write(writer, data);
    if (result != data.len) {
        return error.Writer;
    }
}

pub fn Writer(AnyWriter: type) type {
    // TODO: comptime verify AnyWriter is writer
    return extern struct {
        const Self = @This();

        writer: *const AnyWriter,

        pub fn init(writer: *const AnyWriter) Self {
            return Self{ .writer = writer };
        }

        pub fn interface(self: *Self) InterfaceWriter {
            const writer = lib.GciInterfaceWriter{ .context = self, .write = writeCallback };
            return .{ .writer = writer };
        }

        fn writeCallback(context: ?*const anyopaque, data: [*c]const u8, data_size: usize) callconv(.C) usize {
            std.debug.assert(null != context);
            std.debug.assert(null != data);

            const self: *Self = @constCast(@alignCast(@ptrCast(context)));
            const w: *const AnyWriter = self.writer;
            const d = @as(*[]u8, @constCast(@ptrCast(&.{ .ptr = data, .len = data_size }))).*;
            return w.write(d) catch 0;
        }
    };
}

pub const File = struct {
    inner: lib.GciWriterFile,

    pub fn init(file: *lib.FILE) !File {
        var self: File = undefined;
        const err = lib.gci_writer_file_init(&self.inner, file);
        try internal.enumToError(err);
        return self;
    }

    pub fn interface(self: *File) InterfaceWriter {
        return .{ .writer = lib.gci_writer_file_interface(&self.inner) };
    }
};

pub const String = struct {
    inner: lib.GciWriterString,

    pub fn init(buffer: []u8) !String {
        if (buffer.len > std.math.maxInt(c_int)) {
            return error.Overflow;
        }

        var self: String = undefined;
        const err = lib.gci_writer_string_init(
            &self.inner,
            buffer.ptr,
            @intCast(buffer.len),
        );
        try internal.enumToError(err);
        return self;
    }

    pub fn interface(self: *String) InterfaceWriter {
        return .{ .writer = lib.gci_writer_string_interface(&self.inner) };
    }
};

pub const Buffer = struct {
    inner: lib.GciWriterBuffer,

    pub fn init(writer: InterfaceWriter, buffer: []u8) !Buffer {
        if (buffer.len >= std.math.maxInt(c_int)) {
            return error.Overflow;
        }

        var self: Buffer = undefined;
        const err = lib.gci_writer_buffer_init(
            &self.inner,
            writer.writer,
            buffer.ptr,
            @intCast(buffer.len),
        );
        try internal.enumToError(err);
        return self;
    }

    pub fn interface(self: *Buffer) InterfaceWriter {
        return .{ .writer = lib.gci_writer_buffer_interface(&self.inner) };
    }

    pub fn flush(self: *Buffer) !void {
        const result = lib.gci_writer_buffer_flush(&self.inner);
        if (!result) {
            return error.Writer;
        }
    }
};

const testing = std.testing;
const builtin = @import("builtin");
const clib = @cImport({
    @cInclude("stdio.h");
});

test "c tests" {
    _ = @import("test_writer.zig");
}

test "zig writer" {
    const Fifo = std.fifo.LinearFifo(u8, .Slice);
    const FifoWriter = Writer(Fifo.Writer);

    var buffer: [1]u8 = undefined;
    var fifo = Fifo.init(&buffer);

    var context = FifoWriter.init(&fifo.writer());
    const writer = context.interface();

    try writer.write("1");
    try testing.expectEqualStrings("1", &buffer);
}

test "file init" {
    var context = try File.init(@ptrFromInt(256));
    _ = context.interface();
}

test "file write" {
    var file: [*c]clib.FILE = undefined;

    switch (builtin.os.tag) {
        .linux => {
            file = clib.tmpfile();
        },
        .windows => {
            @compileError("TODO: allow testing file writer, something to do with `GetTempFileNameA` and `GetTempPathA`");
        },
        else => {
            std.debug.print("TODO: allow testing file writer on this os.\n", .{});
            return;
        },
    }
    defer _ = clib.fclose(file);

    var context = try File.init(@as([*c]lib.FILE, @ptrCast(file)));
    const writer = context.interface();

    try writer.write("1");

    const seek_err = clib.fseek(file, 0, lib.SEEK_SET);
    try testing.expectEqual(seek_err, 0);

    var buffer: [2]u8 = undefined;
    const result = clib.fread(&buffer, 1, 2, file);
    try testing.expectEqual(1, result);

    try testing.expectEqualStrings("1", buffer[0..1]);
}

test "string init" {
    var buffer: [0]u8 = undefined;
    var context = try String.init(&buffer);
    _ = context.interface();
}

test "string init overflow" {
    var fake_large_buffer = try testing.allocator.alloc(u8, 2);
    fake_large_buffer.len = @as(usize, std.math.maxInt(c_int)) + 1;
    defer {
        fake_large_buffer.len = 2;
        testing.allocator.free(fake_large_buffer);
    }

    const err = String.init(@ptrCast(fake_large_buffer));
    try testing.expectError(error.Overflow, err);
}

test "string write" {
    var buffer: [2]u8 = undefined;
    var context = try String.init(&buffer);
    const writer = context.interface();

    try writer.write("12");
    try testing.expectEqualStrings("12", &buffer);
}

test "string overflow" {
    var buffer: [0]u8 = undefined;
    var context = try String.init(&buffer);
    const writer = context.interface();

    const err = writer.write("1");
    try testing.expectError(error.Writer, err);
}

test "buffer init" {
    var b: [3]u8 = undefined;
    var c = try String.init(&b);

    var buffer: [1]u8 = undefined;
    var context = try Buffer.init(c.interface(), &buffer);
    _ = context.interface();
}

test "buffer init buffer small" {
    var b: [3]u8 = undefined;
    var c = try String.init(&b);

    var buffer: [0]u8 = undefined;
    const err = Buffer.init(c.interface(), &buffer);
    try testing.expectError(error.Buffer, err);
}

test "buffer init overflow" {
    var fake_large_buffer = try testing.allocator.alloc(u8, 2);
    fake_large_buffer.len = @as(usize, std.math.maxInt(c_int)) + 1;
    defer {
        fake_large_buffer.len = 2;
        testing.allocator.free(fake_large_buffer);
    }

    var b: [3]u8 = undefined;
    var c = try String.init(&b);

    const err = Buffer.init(c.interface(), @ptrCast(fake_large_buffer));
    try testing.expectError(error.Overflow, err);
}

test "buffer write" {
    var b: [1]u8 = undefined;
    var c = try String.init(&b);

    var buffer: [1]u8 = undefined;
    var context = try Buffer.init(c.interface(), &buffer);
    const writer = context.interface();

    try writer.write("1");
    try testing.expectEqualStrings("1", &b);
}

test "buffer flush" {
    var b: [1]u8 = undefined;
    var c = try String.init(&b);

    var buffer: [2]u8 = undefined;
    var context = try Buffer.init(c.interface(), &buffer);
    const writer = context.interface();

    try writer.write("1");

    try context.flush();
    try testing.expectEqualStrings("1", &b);
}

test "buffer internal writer fail" {
    var b: [0]u8 = undefined;
    var c = try String.init(&b);

    var buffer: [1]u8 = undefined;
    var context = try Buffer.init(c.interface(), &buffer);
    const writer = context.interface();

    const err = writer.write("1");
    try testing.expectError(error.Writer, err);
}

test "buffer flush writer fail" {
    var b: [0]u8 = undefined;
    var c = try String.init(&b);

    var buffer: [2]u8 = undefined;
    var context = try Buffer.init(c.interface(), &buffer);
    const writer = context.interface();

    try writer.write("1");

    const err = context.flush();
    try testing.expectError(error.Writer, err);
}

File: src/implementation/writer/writer.c
#include <stdint.h>
#include <limits.h>
#include <string.h>
#include <gci_writer.h>

size_t gci_writer_file_write(void const *void_context, char const *data, size_t data_size);
size_t gci_writer_string_write(void const *void_context, char const *data, size_t data_size);
size_t gci_writer_buffer_write(void const *void_context, char const *data, size_t data_size);

enum GciError gci_writer_file_init(struct GciWriterFile *context, FILE *file) {
    if (context == NULL) { return GCI_ERROR_NULL; }

    context->file = file;
    if (file == NULL) { return GCI_ERROR_NULL; }

    return GCI_ERROR_OK;
}

struct GciInterfaceWriter gci_writer_file_interface(struct GciWriterFile *writer) {
    assert(writer != NULL);
    return (struct GciInterfaceWriter) { .context = writer, .write = gci_writer_file_write };
}

size_t gci_writer_file_write(void const *context, char const *data, size_t data_size) {
    assert(context != NULL);
    assert(data != NULL);
    struct GciWriterFile *writer = (struct GciWriterFile*) context;
    return fwrite(data, sizeof(char), data_size, writer->file);
}

enum GciError gci_writer_string_init(
    struct GciWriterString *context,
    char *buffer,
    size_t buffer_size
) {
    if (context == NULL) { return GCI_ERROR_NULL; }
    context->buffer = buffer;
    if (buffer == NULL) { return GCI_ERROR_NULL; }

    context->buffer_size = buffer_size;
    context->current = 0;

    return GCI_ERROR_OK;
}

struct GciInterfaceWriter gci_writer_string_interface(struct GciWriterString *context) {
    return (struct GciInterfaceWriter) { .context = context, .write = gci_writer_string_write };
}

size_t gci_writer_string_write(void const *void_context, char const *data, size_t data_size) {
    assert(void_context != NULL);
    assert(data != NULL);

    struct GciWriterString *context = (struct GciWriterString*) void_context;
    assert(context->buffer != NULL);
    if (context->buffer_size <= 0) {
        assert(context->current == 0);
    } else {
        assert(0 <= context->current && context->current <= context->buffer_size);
    }

    size_t write_length = context->buffer_size - context->current;
    write_length = write_length > data_size ? data_size : write_length;

    memcpy(context->buffer + context->current, data, write_length);
    context->current += write_length;

    return write_length;
}

enum GciError gci_writer_buffer_init(
        struct GciWriterBuffer *context,
        struct GciInterfaceWriter writer,
        char *buffer,
        size_t buffer_size
) {
    if (context == NULL) { return GCI_ERROR_NULL; }
    context->buffer = buffer;
    if (buffer == NULL) { return GCI_ERROR_NULL; }
    if (buffer_size <= 0) { return GCI_ERROR_BUFFER; }

    context->writer = writer;
    context->buffer_size = buffer_size;
    context->current = 0;

    return GCI_ERROR_OK;
}

struct GciInterfaceWriter gci_writer_buffer_interface(struct GciWriterBuffer *context) {
    return (struct GciInterfaceWriter) { .context = context, .write = gci_writer_buffer_write };
}

size_t gci_writer_buffer_write(void const *void_context, char const *data, size_t data_size) {
    assert(void_context != NULL);
    assert(data != NULL);

    struct GciWriterBuffer *context = (struct GciWriterBuffer*) void_context;
    assert(context->buffer != NULL);
    assert(0 <= context->current && context->current < context->buffer_size);

    size_t write_length = context->buffer_size - context->current;
    write_length = write_length > data_size ? data_size : write_length;

    memcpy(context->buffer + context->current, data, write_length);
    context->current += write_length;

    if (context->current >= context->buffer_size) {
        bool flush_success = gci_writer_buffer_flush(context);
        if (!flush_success) { return 0; }

        if (data_size - write_length >= context->buffer_size) {
            size_t result = gci_writer_write(context->writer, data + write_length, data_size - write_length);
            if (result < data_size - write_length) { return write_length + result; }
        } else {
            memcpy(context->buffer, data + write_length, data_size + write_length);
        }
    }
    return data_size;
}

bool gci_writer_buffer_flush(struct GciWriterBuffer *context) {
    assert(context != NULL);
    assert(context->buffer != NULL);
    assert(0 <= context->current && context->current <= context->buffer_size);

    size_t length = context->current;
    size_t result = gci_writer_write(context->writer, context->buffer, length);
    context->current = 0;

    if (result != length) {
        return false;
    } else {
        return true;
    }
}
